### ISSUE TITLE
tflite op ( attention, or kv update)

### ISSUE THREAD

**Issue by nigelzzz** [google-ai-edge/ai-edge-torch#148]
### Description of the bug:

hi i have transfer tinyllama to tflite format, but when i use `https://netron.app/`, it can show to customer op, can i know the usage in tensorflow.
when don't expand it (e.g., mul, add, softmax...)
![image](https://github.com/user-attachments/assets/fc778648-9eb5-4930-8518-ca5d16ecb817)
![image](https://github.com/user-attachments/assets/1f94e4ff-44c7-49e7-b987-ec4173c262be)

**Comment by pkgoogle** [comment]
Hi @nigelzzz, that is updating the kv_cache so that we do not do extraneous calculations when performing the internal matrix multiplications of the transformer model, here's a good article that explains it: https://medium.com/@joaolages/kv-caching-explained-276520203249 (Also has some information on scaled dot product attention).

**Comment by nigelzzz** [comment]
@pkgoogle ,
  I got it, thanks.
If `scale_dot_product_attention` no has op (mult, soft), how to speed up the caulation, (maybe use npu's op(mult...))

**Comment by nigelzzz** [comment]
Hi @pkgoogle ,
       "In my understanding, scale_dot_product_attention is equivalent to a combination of matmul, softmax, mask, and scale operations. If I want to break down scale_dot_product_attention into these four operations, how can I modify the code?"
  
![image](https://github.com/user-attachments/assets/fb9bf760-8330-4564-b88a-ced74d515f7e)

**Comment by pkgoogle** [comment]
Hi @nigelzzz, I would read go-through the Generative API: https://github.com/google-ai-edge/ai-edge-torch/tree/main/ai_edge_torch/generative, Do one example where you just reauthor the basic transformer block rather than changing anything for now. Make sure that works first. Review the toy_model: https://github.com/google-ai-edge/ai-edge-torch/blob/main/ai_edge_torch/generative/examples/test_models/toy_model.py. In the forward method is where you'll find how it's defined on the forward pass, you'll see it has a Transformer block. You'll find it defined here: https://github.com/google-ai-edge/ai-edge-torch/blob/main/ai_edge_torch/generative/layers/attention.py. Attention has numerous different forms, so it may not look exactly the same as the original paper. But keep digging and you'll find something similar.
